Extension/src/LanguageServer/extension.ts

```diff
@@ -60,6 +62,68 @@ export function activate(activationEventOccurred: boolean): void {
     }
 }
 
+export function registerCustomConfigurationProvider(provider: CustomConfigurationProvider): void {
+    customConfigurationProviders.push(provider);
+}
+
+export async function provideCustomConfiguration(document: vscode.TextDocument, client: Client): Promise<void> {
+    return runBlockingThenableWithTimeout(async () => {
+        // Loop through registered providers until one is able to service the current document
+        for (let i: number = 0; i < customConfigurationProviders.length; i++) {
+            if (await customConfigurationProviders[i].canProvideConfiguration(document.uri)) {
+                return customConfigurationProviders[i].provideConfigurations([document.uri]);
+            }
+        }
+        return Promise.reject("No providers found for " + document.uri);
+    }, 1000, client).then((configs: SourceFileConfigurationItem[]) => {
+        if (configs !== null && configs.length > 0) {
+            client.sendCustomConfigurations(configs);
+        }
+    });
+}
+
+export function onDidCustomConfigurationChange(customConfigurationProvider: CustomConfigurationProvider): void {
+    // TODO: How to handle multiple workspaces? Only works with the current workspace.
+    // Should this be called when the workspace changes?
+    if (vscode.workspace.textDocuments === undefined || vscode.workspace.textDocuments.length === 0) {
+        return;
+    }
+
+    let documents: vscode.Uri[] = [];
+    for (let i: number = 0; i < vscode.workspace.textDocuments.length; ++i) {
+        let document: vscode.TextDocument = vscode.workspace.textDocuments[i];
+        if (document.languageId === "cpp" || document.languageId === "c") {
+            documents.push(document.uri);
+        }
+    }
+
+    // TODO: Can we rely on clients.ActiveClient to be the client for the current workspace?
+    runBlockingThenableWithTimeout(() => customConfigurationProvider.provideConfigurations(documents), 1000, clients.ActiveClient)
+    .then((configs: SourceFileConfigurationItem[]) => {
+        clients.ActiveClient.sendCustomConfigurations(configs);
+    });
+}
+
+function runBlockingThenableWithTimeout(thenable: () => Thenable<any>, ms: number, client: Client): Thenable<any> {
+    let timer: NodeJS.Timer;
+    // Create a promise that rejects in <ms> milliseconds
+    let timeout: Promise<any> = new Promise((resolve, reject) => {
+        timer = setTimeout(() => {
+            clearTimeout(timer);
+            reject("Timed out in " + ms + "ms.");
+        }, ms);
+    });
+
+    // Returns a race between our timeout and the passed in promise
+    return client.runBlockingTask(Promise.race([thenable(), timeout]).then((result: any) => {
+        clearTimeout(timer);
```

Comment:
This line is unnecessary since the timer is cleared in the timeout promise. Will remove in the next push.